### PR TITLE
Laboratory work #1, Dasha Oleinik - 23FPL1

### PR DESCRIPTION
**Pull request by dkole13**

### CODE REVIEW

File: lab_1_classify_by_unigrams/start.py [hunk 1]
```diff
@@ -1,20 +1,26 @@
...
+        'assets/profiles/de.json',
+        'assets/profiles/en.json',
+        'assets/profiles/es.json',
+        'assets/profiles/fr.json',
+        'assets/profiles/it.json',
+        'assets/profiles/ru.json',
+        'assets/profiles/tr.json'
+    ]
+
+    result = 'AAAAA'
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:23]: ok

**Review by EIzharikov — CHANGES_REQUESTED** [review]
There are problems with removing docstrings and indentation, and a few questions in the code, but in general the work is good! I see that you tried hard. I wish you good luck in further laboratory work!

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -35,6 +49,21 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
     In case of corrupt input arguments, None is returned
     """
 
+    # checking input
+    if not isinstance(tokens, list) or not all(isinstance(item, str) for item in tokens):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:53]: what happens if an empty list arrives?

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -35,6 +49,21 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+        return None
+
+    tokens_total = len(tokens)
+
+    freq = {}
+    for char in set(tokens):
+        freq[char] = float(tokens.count(char))
+
+    for letter in freq:
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:62]: can you use only one for loop?

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -50,6 +79,14 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
     In case of corrupt input arguments, None is returned
     """
 
+    # checking input
+    if not isinstance(language, str) or not isinstance(text, str):
+        return None
+
+    profile = {'name': language, 'freq': calculate_frequencies(tokenize(text))}
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:86]: what will happen if calculate_frequencies will return None?

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -1,140 +1,133 @@
...
-    """
+    return round(mse, 4)
 
 
 def compare_profiles(
     unknown_profile: dict[str, str | dict[str, float]],
     profile_to_compare: dict[str, str | dict[str, float]],
-) -> float | None:
-    """
-    Compare profiles and calculate the distance using symbols.
```
[the remark this answers is not in the document]
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py:74]: rm indents

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -85,12 +129,31 @@ def compare_profiles(
     'freq' in arguments, None is returned
     """
 
+    # checking input
+    if not isinstance(unknown_profile, dict) or not isinstance(profile_to_compare, dict):
+        return None
+    if 'name' not in unknown_profile.keys() or 'name' not in profile_to_compare.keys():
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:135]: you can use only unkwon_profiel without .keys()

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -120,6 +205,15 @@ def load_profile(path_to_file: str) -> dict | None:
     In case of corrupt input arguments, None is returned
     """
 
+    # checking input
+    if not isinstance(path_to_file, str):
+        return None
+
+    with open(path_to_file, 'r', encoding='UTF-8') as file:
+        profile = dict(load(file))
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:213]: why are you converting to dict?

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -150,10 +264,19 @@ def collect_profiles(paths_to_profiles: list) -> list[dict[str, str | dict[str,
     In case of corrupt input arguments, None is returned
     """
 
+    # checking input
+    if not isinstance(paths_to_profiles, list):
+        return None
+
+    loaded_profiles = []
+    for path in paths_to_profiles:
+        loaded_profiles.append(preprocess_profile(load_profile(path)))
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:273]: what if any function  return None?

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -168,6 +291,18 @@ def detect_language_advanced(
...
 
+    # checking input
+    if not isinstance(unknown_profile, dict) or not isinstance(known_profiles, list):
+        return None
+
+    distances = []
+    for profile in known_profiles:
+        distances.append(tuple((profile['name'], compare_profiles(unknown_profile, profile))))
+
+    # distances = sorted(distances, key=itemgetter(1, 0))
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:302]: delete

**Review by EIzharikov — COMMENTED** [review]
Nice work, but MyPy does not think so. You can achieve 10, I believe in you!

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 6 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 7 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: removed

File: lab_1_classify_by_unigrams/main.py (hunk 8 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 9 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -152,7 +145,7 @@ def collect_profiles(paths_to_profiles: list) -> list[dict[str, str | dict[str,
 
 
 def detect_language_advanced(
-    unknown_profile: dict[str, str | dict[str, float]], known_profiles: list
+        unknown_profile: dict[str, str | dict[str, float]], known_profiles: list
```
[the remark this answers is not in the document]
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py]: ??

File: lab_1_classify_by_unigrams/main.py [hunk 11]
```diff
@@ -50,6 +77,20 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
...
+    if not isinstance(language, str) or not isinstance(text, str):
+        return None
+
+    frequencies = calculate_frequencies(tokenize(text))
+    if (not isinstance(frequencies, dict)
+            or not all(isinstance(char, str) for char in frequencies)
+            or not all(isinstance(value, float) for value in frequencies.values())):
+        return None
+
+    if frequencies is not None:
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:90]: may be you can swap the conditions, make here early stop

File: lab_1_classify_by_unigrams/main.py [hunk 12]
```diff
@@ -65,11 +106,18 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
     In case of corrupt input arguments, None is returned
     """
 
+    # checking input
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:109]: delete comments, in laboratory works they are excess

File: lab_1_classify_by_unigrams/main.py [hunk 13]
```diff
@@ -65,11 +106,18 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
+
+    return float(sum((actual[i] - predicted[i]) ** 2
+                     for i in range(len(actual))) / len(actual))
+
 
 def compare_profiles(
-    unknown_profile: dict[str, str | dict[str, float]],
-    profile_to_compare: dict[str, str | dict[str, float]],
-) -> float | None:
+        unknown_profile: dict[str, str | dict[str, float]],
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:119]: revert these changes

File: lab_1_classify_by_unigrams/main.py [hunk 14]
```diff
@@ -85,12 +133,37 @@ def compare_profiles(
     'freq' in arguments, None is returned
     """
 
+    # checking input
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:136]: .

File: lab_1_classify_by_unigrams/main.py [hunk 15]
```diff
@@ -85,12 +133,37 @@ def compare_profiles(
...
+    if not isinstance(unknown_profile, dict) or not isinstance(profile_to_compare, dict):
+        return None
+    if 'name' not in unknown_profile or 'name' not in profile_to_compare:
+        return None
+    if not all(isinstance(key, str) for key in unknown_profile) \
+            or not all(isinstance(key, str) for key in profile_to_compare):
+        return None
+
+    all_chars = set(unknown_profile['freq']).union(set(profile_to_compare['freq']))
+    # creating lists to pass into calculate_mse function
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:146]: .

File: lab_1_classify_by_unigrams/main.py [hunk 16]
```diff
@@ -85,12 +133,37 @@ def compare_profiles(
...
+    unknown_profile_freq = unknown_profile['freq']
+    if not isinstance(unknown_profile_freq, dict):
+        return None
+    unknown_sorted = [(unknown_profile_freq[char] if char in unknown_profile_freq else 0.0)
+                      for char in all_chars]
+
+    profile_to_compare_freq = profile_to_compare['freq']
+    if not isinstance(profile_to_compare_freq, dict):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:154]: can we check it earlier?

File: lab_1_classify_by_unigrams/main.py [hunk 17]
```diff
@@ -136,6 +243,26 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
     'freq' in arguments, None is returned
     """
 
+    # checking input
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:246]: .

File: lab_1_classify_by_unigrams/main.py [hunk 18]
```diff
@@ -150,10 +277,25 @@ def collect_profiles(paths_to_profiles: list) -> list[dict[str, str | dict[str,
     In case of corrupt input arguments, None is returned
     """
 
+    # checking input
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:280]: .

File: lab_1_classify_by_unigrams/main.py [hunk 19]
```diff
@@ -150,10 +277,25 @@ def collect_profiles(paths_to_profiles: list) -> list[dict[str, str | dict[str,
...
+        else:
+            return None
+
+    return loaded_profiles
+
 
 def detect_language_advanced(
-    unknown_profile: dict[str, str | dict[str, float]], known_profiles: list
-) -> list | None:
+        unknown_profile: dict[str, str | dict[str, float]], known_profiles: list) -> list | None:
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:298]: revert indents

File: lab_1_classify_by_unigrams/main.py [hunk 20]
```diff
@@ -168,6 +310,20 @@ def detect_language_advanced(
     In case of corrupt input arguments, None is returned
     """
 
+    # checking input
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:313]: .

File: lab_1_classify_by_unigrams/main.py [hunk 21]
```diff
@@ -178,3 +334,12 @@ def print_report(detections: list[tuple[str, float]]) -> None:
 
     In case of corrupt input arguments, None is returned
     """
+
+    # checking input
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:338]: .

File: lab_1_classify_by_unigrams/settings.json [hunk 22]
```diff
@@ -1,3 +1,3 @@
 {
-  "target_score": 0
+  "target_score": 10
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/settings.json:2]: NICE!!!

File: lab_1_classify_by_unigrams/start.py [hunk 23]
```diff
@@ -3,18 +3,47 @@
...
+
+    profiles = collect_profiles(paths_to_profiles)
+    unknown_profile = create_language_profile('unknown', unknown_text)
+
+    if (not isinstance(profiles, list)
+            or not isinstance(unknown_profile, dict)):
+        result = None
+    else:
+        detections = detect_language_advanced(unknown_profile, profiles)
+        if (not isinstance(detections, list)
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:40]: you are not believe in your functions?

File: lab_1_classify_by_unigrams/start.py [hunk 24]
```diff
@@ -3,18 +3,47 @@
...
+        'assets/profiles/fr.json',
+        'assets/profiles/it.json',
+        'assets/profiles/ru.json',
+        'assets/profiles/tr.json'
+    ]
+
+    profiles = collect_profiles(paths_to_profiles)
+    unknown_profile = create_language_profile('unknown', unknown_text)
+
+    if (not isinstance(profiles, list)
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:35]: you are not believe in your functions?

File: seminars/practice_2_strings.py [hunk 25]
```diff
@@ -5,196 +5,3 @@
 
 Data Type: String
 """
-
```
Comment (EIzharikov) [inline seminars/practice_2_strings.py:8]: revert all changes in seminars

**Review by EIzharikov — COMMENTED** [review]
Nice progress! I am very glad that you are aimed to 10! Fix my comments, make CI green and be ready for the next review! Good luck!

File: lab_1_classify_by_unigrams/main.py (hunk 11 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 12 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: all deleted now

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py:113]: done

File: lab_1_classify_by_unigrams/main.py (hunk 14 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 15 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 16 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 17 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 18 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 19 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 20 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 21 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/settings.json (hunk 22 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/settings.json:2]: THANK YOU!!!!

File: lab_1_classify_by_unigrams/start.py (hunk 23 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/start.py:40]: i do, mypy does not :(

File: seminars/practice_2_strings.py (hunk 25 above)
Comment (dkole13) [reply seminars/practice_2_strings.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 13 above)
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py:113]: i do not see any changes here

File: lab_1_classify_by_unigrams/main.py [hunk 26]
```diff
@@ -85,12 +128,34 @@ def compare_profiles(
...
+        return None
+    if not all(isinstance(key, str) for key in unknown_profile) \
+            or not all(isinstance(key, str) for key in profile_to_compare) \
+            or not isinstance(unknown_profile['freq'], dict) \
+            or not isinstance(profile_to_compare['freq'], dict):
+        return None
+
+    all_chars = set(unknown_profile['freq']).union(set(profile_to_compare['freq']))
+
+    unknown_profile_freq = unknown_profile['freq']
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:143]: u can create this variable before the creating `all_chars` list and then use this variables in `all_chars` list creating

File: lab_1_classify_by_unigrams/main.py [hunk 27]
```diff
@@ -1,140 +1,133 @@
...
+    mse = calculate_mse(unknown_sorted, list(profile_to_compare['freq'].values()))
+
+    return mse
 
 
 def detect_language(
-    unknown_profile: dict[str, str | dict[str, float]],
-    profile_1: dict[str, str | dict[str, float]],
-    profile_2: dict[str, str | dict[str, float]],
+        unknown_profile: dict[str, str | dict[str, float]],
```
[the remark this answers is not in the document]
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py:156]: it is actual

File: lab_1_classify_by_unigrams/main.py [hunk 28]
```diff
@@ -150,9 +264,24 @@ def collect_profiles(paths_to_profiles: list) -> list[dict[str, str | dict[str,
     In case of corrupt input arguments, None is returned
     """
 
+    if not isinstance(paths_to_profiles, list):
+        return None
+
+    loaded_profiles = []
+    for path in paths_to_profiles:
+        profile_loaded = load_profile(path)
+        if isinstance(profile_loaded, dict):
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:273]: can we make here early stop ?

File: lab_1_classify_by_unigrams/main.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py:284]: ??

File: lab_1_classify_by_unigrams/main.py [hunk 29]
```diff
@@ -168,6 +297,19 @@ def detect_language_advanced(
...
     """
 
+    if not isinstance(unknown_profile, dict) or not isinstance(known_profiles, list):
+        return None
+
+    distances = []
+    for profile in known_profiles:
+        profile_name: str = profile['name']
+        distance = compare_profiles(unknown_profile, profile)
+        if distance is None:
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:307]: you can replace this by `if not distance`

Comment (EIzharikov) [reply lab_1_classify_by_unigrams/main.py:307]: or not?

File: lab_1_classify_by_unigrams/main.py [hunk 30]
```diff
@@ -168,6 +297,19 @@ def detect_language_advanced(
...
+            return None
+        distances.append((profile_name, distance))
+    distances_sorted: list[tuple['str', float]] = sorted(distances, key=lambda t: t[1])
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/main.py:310]: can you rename `t`?

File: lab_1_classify_by_unigrams/start.py [hunk 31]
```diff
@@ -3,18 +3,47 @@
...
 
+from lab_1_classify_by_unigrams.main import (collect_profiles, create_language_profile,
+                                             detect_language_advanced, print_report)
+
 
 def main() -> None:
     """
     Launches an implementation
     """
+
```
Comment (EIzharikov) [inline lab_1_classify_by_unigrams/start.py:14]: delete whitespace

**Review by EIzharikov — COMMENTED** [review]
Great job, well done! Correct the code according to my comments, make sure the CI is green and get ready to defend the lab work! Check previous coments!*

File: lab_1_classify_by_unigrams/main.py (hunk 26 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py:141]: done

File: lab_1_classify_by_unigrams/main.py (hunk 28 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 29 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: done

File: lab_1_classify_by_unigrams/main.py (hunk 30 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/main.py]: renamed

File: lab_1_classify_by_unigrams/start.py (hunk 31 above)
Comment (dkole13) [reply lab_1_classify_by_unigrams/start.py]: done
